pkg/controllers/clusterresourcequota/clusterresourcequota_controller.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterresourcequota
+
+import (
+	"context"
+
+	platformv1 "github.com/flanksource/platform-operator/pkg/apis/platform/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilquota "k8s.io/kubernetes/pkg/quota/v1"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const name = "clusterresourcequota-controller"
+
+var log = logf.Log.WithName(name)
+
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileClusterResourceQuota{
+		Client: mgr.GetClient(),
+		Scheme: mgr.GetScheme(),
+	}
+}
+
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	c, err := controller.New(name, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &platformv1.ClusterResourceQuota{}}, &handler.EnqueueRequestForObject{},
+	); err != nil {
+		return err
+	}
+
+	if err := c.Watch(
+		// FIXME: When a new Resource Quota is created the controller does not get reconciled
+		//
+		&source.Kind{Type: &corev1.ResourceQuota{}}, &handler.EnqueueRequestForOwner{
```

Comment:
This logic is broken.

We would expect that in case of an Event generated by the `ResourceQuota` object, we would reconcile the request and update the status of the `ClusterResourceQuota` object.

However, from the tests I have seen that this does not work because actually Resource Quotas are not owned by ClusterResourceQuota.

In order to do so, we would need to build our own custom handler and pass it here so when a ResourceQuota gets changed in any namespace in the cluster, the ClusterResourceQuota request gt queued and we have a reconciliation happening.

@moshloop

Reply:
Evaluating if we can achieve the same using `handler.EnqueueRequestsFromMapFunc`

Reply:
`handler.EnqueueRequestsFromMapFunc` made the job. we are able to map any kind of events on `ResourceQuotas` across the entire cluster to a `ClusterResourceQuota`.